go/logic/migrator.go

```diff
@@ -1279,36 +1279,43 @@ func (this *Migrator) executeWriteFuncs() error {
 					return err
 				}
 			}
-		default:
+		case copyRowsFunc := <-this.copyRowsQueue:
+			if err := this.drainApplierEventQueue(); err != nil {
+				return this.migrationContext.Log.Errore(err)
+			}
 			{
-				select {
-				case copyRowsFunc := <-this.copyRowsQueue:
-					{
-						copyRowsStartTime := time.Now()
-						// Retries are handled within the copyRowsFunc
-						if err := copyRowsFunc(); err != nil {
-							return this.migrationContext.Log.Errore(err)
-						}
-						if niceRatio := this.migrationContext.GetNiceRatio(); niceRatio > 0 {
-							copyRowsDuration := time.Since(copyRowsStartTime)
-							sleepTimeNanosecondFloat64 := niceRatio * float64(copyRowsDuration.Nanoseconds())
-							sleepTime := time.Duration(int64(sleepTimeNanosecondFloat64)) * time.Nanosecond
-							time.Sleep(sleepTime)
-						}
-					}
-				default:
-					{
-						// Hmmmmm... nothing in the queue; no events, but also no row copy.
-						// This is possible upon load. Let's just sleep it over.
-						this.migrationContext.Log.Debugf("Getting nothing in the write queue. Sleeping...")
-						time.Sleep(time.Second)
-					}
+				copyRowsStartTime := time.Now()
+				// Retries are handled within the copyRowsFunc
+				if err := copyRowsFunc(); err != nil {
+					return this.migrationContext.Log.Errore(err)
+				}
+				if niceRatio := this.migrationContext.GetNiceRatio(); niceRatio > 0 {
+					copyRowsDuration := time.Since(copyRowsStartTime)
+					sleepTimeNanosecondFloat64 := niceRatio * float64(copyRowsDuration.Nanoseconds())
+					sleepTime := time.Duration(sleepTimeNanosecondFloat64) * time.Nanosecond
+					time.Sleep(sleepTime)
 				}
 			}
 		}
 	}
 }
 
+func (this *Migrator) drainApplierEventQueue() error {
+	for {
+		this.throttler.throttle(nil)
+		select {
+		case eventStruct := <-this.applyEventsQueue:
+			{
+				if err := this.onApplyEventStruct(eventStruct); err != nil {
```

Comment:
```suggestion
	for {
		select {
		case eventStruct := <-this.applyEventsQueue:
			{
				this.throttler.throttle(nil)
				if err := this.onApplyEventStruct(eventStruct); err != nil {
```
Let's only call the throttler if there is an event in the queue

Reply:
The call way I am reference [Line 1271](https://github.com/github/gh-ost/blob/b209349e7fdf16d2823e3ef02c3fbef1fd51598a/go/logic/migrator.go#L1271), the throttler is called before there is an event in the queue

Reply:
@debug-LiXiwen yes, but the `for` loop in `.drainApplierEventQueue()` could potentially write many events

Currently the throttler is checked between every apply event, so I believe `.drainApplierEventQueue()` should do the same on each iteration of the `for` loop on [line 1304](https://github.com/github/gh-ost/blob/b209349e7fdf16d2823e3ef02c3fbef1fd51598a/go/logic/migrator.go#L1304)

[Your last commit](https://github.com/github/gh-ost/commit/b209349e7fdf16d2823e3ef02c3fbef1fd51598a) is already doing this [on line 1305](https://github.com/github/gh-ost/blob/b209349e7fdf16d2823e3ef02c3fbef1fd51598a/go/logic/migrator.go#L1305) but the throttler check is ran before we know if there is anything in the applier queue. I believe the throttler check should occur [on line 1309](https://github.com/github/gh-ost/blob/b209349e7fdf16d2823e3ef02c3fbef1fd51598a/go/logic/migrator.go#L1309) instead - only when there is events to apply

Reply:
oh, you're right,I will modify it